### PR TITLE
Reduce identify calls

### PR DESCRIPTION
**Pull request by ewoutp**
This change reduce the number of identify calls made by the cli in order to improve response times.

### CODE REVIEW

File: delete.go [hunk 1]
```diff
@@ -4,7 +4,7 @@ import (
 	"fmt"
```
Comment (ewoutp) [inline delete.go]: Changes in this file are cleanup only

**Comment by ewoutp** [comment]
@denderello RFR

File: ids/partial_ids.go [hunk 2]
```diff
@@ -42,26 +42,37 @@ type IdArgumentOptions struct {
...
+			// a service name. If this is not true, it will result
+			// in service not found error later on which is no problem.
+			return &IdArgument{Kind: Service, Service: arg}, nil
 		}
 
 		if options.AllowInstanceId {
-			if identity.Kind == apiclient.KindInstance {
-				return &IdArgument{Kind: InstanceId, InstanceId: arg}, nil
+			// The argument is allowed to be a partial instance id, so we must check
+			// first what it is.
```
Comment (zyndiecate) [inline ids/partial_ids.go]: Here we only can check if it is a service name or a proper instance ID. In case it is a partial instance ID, need to go on in later steps. This is not really reflected in the comment.

File: ids/partial_ids.go [hunk 3]
```diff
@@ -42,26 +42,37 @@ type IdArgumentOptions struct {
...
+					} else {
+						// Argument is a service which is not allowed.
+						return nil, maskAny(InvalidArgumentError)
+					}
+				case apiclient.KindInstance:
+					return &IdArgument{Kind: InstanceId, InstanceId: arg}, nil
+				default:
+					// We don't recognize the argument, will fall through
+					// here and check for partial instance id later.
+				}
```
Comment (zyndiecate) [inline ids/partial_ids.go]: I am sure this works, I just think that the code flow is more complicated than it should be. We have a set of logic options. So lets just check them straight away. See pseudo code. We don't need if conditions inside switches, which are inside if conditions. :) 


```
if appAllowed && !instanceAllowed {
  return kind service
}

identity, err := apiclient.FetchIdentity()

if appAllowed && instanceAllowed {
  if err == nil {
    return identity kind
  }
}

if !appAllowed && instanceAllowed {
  // do stuff for partial IDs
}

if !appAllowed && !instanceAllowed {
  // do stuff for components
}
```

File: ids/partial_ids.go [hunk 4]
```diff
@@ -42,26 +42,37 @@ type IdArgumentOptions struct {
...
-			if identity.Kind == apiclient.KindApp {
-				return &IdArgument{Kind: Service, Service: arg}, nil
-			}
+		// Argument consists of one part.
+		// It can be a service name, a full instance id or a partial instance id.
+		if options.AllowService && !options.AllowInstanceId {
+			// Since instance id's are not allowed, we assume it is
+			// a service name. If this is not true, it will result
+			// in service not found error later on which is no problem.
+			return &IdArgument{Kind: Service, Service: arg}, nil
```
Comment (zyndiecate) [inline ids/partial_ids.go]: This here is very nice!

File: stats.go [hunk 5]
```diff
@@ -88,28 +88,41 @@ func processStats(arg string, recursive bool) {
...
-		status, err := client.ComponentStatus(idArg.Service, idArg.Component, recursive)
-		if err != nil {
+		processComponentStats(idArg.Service, idArg.Component, recursive)
+	}
+}
+
+func processServiceStats(service string) {
+	status, err := client.AppStatus(service)
+	if err != nil {
+		Quitf(defaultErrorMsgTemplate, err)
```
Comment (zyndiecate) [inline stats.go]: Nice cleanup, but I think errors should be returned to the caller and handled there.

File: ids/partial_ids.go (hunk 2 above)
Comment (ewoutp) [inline ids/partial_ids.go]: Update the comment

File: stats.go (hunk 5 above)
Comment (ewoutp) [inline stats.go]: This is the general pattern that is being used in the cli.
I agree with you that this should change, but I want to do that in another PR

File: ids/partial_ids.go (hunk 3 above)
Comment (denderello) [inline ids/partial_ids.go]: I like this approach.

Comment (ewoutp) [inline ids/partial_ids.go]: I've simplified the logic. Please have another look

File: ids/partial_ids.go [hunk 6]
```diff
@@ -42,52 +42,59 @@ type IdArgumentOptions struct {
...
+					// Argument is a service which is not allowed.
+					return nil, maskAny(InvalidArgumentError)
+				}
+			case apiclient.KindInstance:
+				return &IdArgument{Kind: InstanceId, InstanceId: arg}, nil
+			default:
+				// We don't recognize the argument, will fall through
+				// here and check for partial instance id later.
+			}
+		}
```
Comment (zyndiecate) [inline ids/partial_ids.go]: Mh, I see no difference to the version before. At this point, this is not important enough to me to further bother you with this. Go ahead.
